🐛 Bug: Can't be built in a Vercel deployment environment

### Bug Report Checklist

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have pulled the latest `main` branch of the repository.
- [X] I have [searched for related issues](https://github.com/JoshuaKGoldberg/tidelift-me-up/issues?q=is%3Aissue) and found none that matched my issue.

### Expected

You should be able to `import { tideliftMeUp } from "tidelift-me-up"` in any reasonable Node environment. For example, an Astro/Next.js/Remix/etc. site deployed on Netflify/Vercel/similar should be able to use this package.

### Actual

In a Next.js app on Vercel (https://github.com/JoshuaKGoldberg/tidelift-me-up-site/tree/46edc10a74add5de79c27e0765ae22edfbb6f689):

```plaintext
- info Creating an optimized production build...
--
17:55:38.658 | Failed to compile.
17:55:38.658 |  
17:55:38.658 | ./node_modules/.pnpm/got@7.1.0/node_modules/got/index.js
17:55:38.658 | Module not found: Can't resolve 'electron'
17:55:38.658 |  
17:55:38.659 | https://nextjs.org/docs/messages/module-not-found
17:55:38.660 |  
17:55:38.661 | Import trace for requested module:
17:55:38.661 | ./node_modules/.pnpm/npm-user-packages@3.0.0/node_modules/npm-user-packages/index.js
17:55:38.661 | ./node_modules/.pnpm/tidelift-me-up@0.1.4/node_modules/tidelift-me-up/lib/tideliftMeUp.js
17:55:38.661 | ./node_modules/.pnpm/tidelift-me-up@0.1.4/node_modules/tidelift-me-up/lib/index.js
17:55:38.661 | ./src/app/page.tsx
```


### Additional Info

`got` is a nested dependency of `npm-user-packages`: https://github.com/kevva/npm-user-packages/blob/aa001c2701ff59f15b6570578c5c27be8363171d/index.js#L2

We should file an issue on them to look into using native `fetch` now that it's available & supported everywhere. If they're not responsive we can always re-implement the same kind of API call here. There's not much code in play.

--- a/README.md
+++ b/README.md
@@ -54,7 +54,7 @@ npx tidelift-me-up
 - `--since` _(default: 2 years ago)_: A date that packages need to have been updated since to be considered
   - This will be provided as a string to the `Date` constructor
 - `--username` _(default: result of `npm whoami`)_: The npm username to search for packages maintained by
-  - The search is done by [`npm-user-packages`](https://github.com/kevva/npm-user-packages), which fetches from [npm.io](https://npm.io)
+  - The search is done by a network call to [npms.io](https://npms.io)
 
 ```shell
 npx tidelift-me-up --ownership author --ownership publisher --reporter json --since 2020 --username your-username

--- a/cspell.json
+++ b/cspell.json
@@ -20,6 +20,7 @@
 		"lcov",
 		"markdownlintignore",
 		"npmignore",
+		"npms",
 		"npmpackagejsonlintrc",
 		"outro",
 		"packagejson",

--- a/package.json
+++ b/package.json
@@ -40,13 +40,11 @@
 	},
 	"dependencies": {
 		"chalk": "^5.2.0",
-		"npm-user-packages": "^3.0.0",
 		"npm-whoami": "^1.1.4"
 	},
 	"devDependencies": {
 		"@types/eslint": "^8.37.0",
 		"@types/node": "^20.3.1",
-		"@types/npm-user-packages": "^3.0.1",
 		"@typescript-eslint/eslint-plugin": "^5.59.5",
 		"@typescript-eslint/parser": "^5.59.5",
 		"@vitest/coverage-istanbul": "^0.31.0",

--- a/package.json
+++ b/package.json
@@ -81,7 +81,7 @@
 		"vitest": "^0.32.0",
 		"yaml-eslint-parser": "^1.2.2"
 	},
-	"packageManager": "pnpm@8.6.0",
+	"packageManager": "pnpm@8.6.2",
 	"engines": {
 		"node": ">=18"
 	},

--- a/src/createUserPackagesFilter.ts
+++ b/src/createUserPackagesFilter.ts
@@ -1,5 +1,4 @@
-import { PackageData } from "npm-user-packages";
-
+import { PackageData } from "./getNpmUserPackages.js";
 import { PackageOwnership } from "./types.js";
 
 export interface FilterSettings {

--- a/src/getNpmUserPackages.test.ts
+++ b/src/getNpmUserPackages.test.ts
@@ -23,15 +23,22 @@ describe("getNpmUserPackages", () => {
 	});
 
 	it("returns results when they are paginated", async () => {
-		const packageNames = new Array(602).fill(undefined).map((_, i) => `${i}`);
+		const total = 501;
+		const packageNames = new Array(total).fill(undefined).map((_, i) => `${i}`);
 
-		mockFetch.mockResolvedValue({
-			json: () => ({
-				results: packageNames.map((responseName) => ({
-					package: responseName,
-				})),
-				total: 2,
-			}),
+		let calls = 0;
+		mockFetch.mockImplementation(() => {
+			const start = calls * 250;
+			const end = (calls + 1) * 250;
+			calls += 1;
+			return {
+				json: () => ({
+					results: packageNames.slice(start, end).map((responseName) => ({
+						package: responseName,
+					})),
+					total,
+				}),
+			};
 		});
 
 		const actual = await getNpmUserPackages("maintainer");

--- a/src/getNpmUserPackages.test.ts
+++ b/src/getNpmUserPackages.test.ts
@@ -1,0 +1,41 @@
+import { describe, expect, it, vi } from "vitest";
+
+import { getNpmUserPackages } from "./getNpmUserPackages.js";
+
+const mockFetch = vi.fn();
+
+vi.stubGlobal("fetch", mockFetch);
+
+describe("getNpmUserPackages", () => {
+	it("returns results when they are not paginated", async () => {
+		const packageNames = ["abc", "def", "ghi"];
+
+		mockFetch.mockResolvedValue({
+			json: () => ({
+				results: packageNames.map((packageName) => ({ package: packageName })),
+				total: 2,
+			}),
+		});
+
+		const actual = await getNpmUserPackages("maintainer");
+
+		expect(actual).toEqual(packageNames);
+	});
+
+	it("returns results when they are paginated", async () => {
+		const packageNames = new Array(602).fill(undefined).map((_, i) => `${i}`);
+
+		mockFetch.mockResolvedValue({
+			json: () => ({
+				results: packageNames.map((responseName) => ({
+					package: responseName,
+				})),
+				total: 2,
+			}),
+		});
+
+		const actual = await getNpmUserPackages("maintainer");
+
+		expect(actual).toEqual(packageNames);
+	});
+});

--- a/src/getNpmUserPackages.ts
+++ b/src/getNpmUserPackages.ts
@@ -1,0 +1,71 @@
+/*
+This is a re-implementation of the `npm-user-packages` package API for:
+* https://github.com/JoshuaKGoldberg/tidelift-me-up/issues/26
+* https://github.com/kevva/npm-user-packages/issues/1
+
+If you still see this in August 2023 or later, it means that issue is stale
+(or you need to pull the latest tidelift-me-up code/version).
+In which case I (Josh) will want to publish my own get-npm-user-packages...
+*/
+
+export interface PackageData {
+	author: {
+		email?: string | undefined;
+		name: string;
+		url?: string | undefined;
+		username?: string | undefined;
+	};
+	date: string;
+	description: string;
+	keywords: string[];
+	links: {
+		bugs?: string | undefined;
+		homepage?: string | undefined;
+		npm: string;
+		repository?: string | undefined;
+	};
+	maintainers: {
+		email: string;
+		username: string;
+	}[];
+	name: string;
+	publisher: {
+		email: string;
+		username: string;
+	};
+	scope: string;
+	version: string;
+}
+
+interface ResponseBody {
+	results: ResponseResult[];
+	total: number;
+}
+
+interface ResponseResult {
+	package: PackageData;
+}
+
+export async function getNpmUserPackages(maintainer: string) {
+	const packages: PackageData[] = [];
+	const size = 250;
+	let from = 0;
+
+	while (true) {
+		const response = await fetch(
+			`https://api.npm.io/v2/search?q=maintainer:${maintainer}&size=${size}&from=${from}`,
+			{
+				headers: { "Content-Type": "application/json" },
+			}
+		);
+		const body = (await response.json()) as ResponseBody;
+
+		packages.push(...body.results.map((result) => result.package));
+
+		if (body.total <= from || body.total > size) {
+			from += size - 1;
+		} else {
+			return packages;
+		}
+	}
+}

--- a/src/getNpmUserPackages.ts
+++ b/src/getNpmUserPackages.ts
@@ -49,11 +49,10 @@ interface ResponseResult {
 export async function getNpmUserPackages(maintainer: string) {
 	const packages: PackageData[] = [];
 	const size = 250;
-	let from = 0;
 
 	while (true) {
 		const response = await fetch(
-			`https://api.npm.io/v2/search?q=maintainer:${maintainer}&size=${size}&from=${from}`,
+			`https://api.npm.io/v2/search?q=maintainer:${maintainer}&size=${size}&from=${packages.length}`,
 			{
 				headers: { "Content-Type": "application/json" },
 			}
@@ -62,9 +61,7 @@ export async function getNpmUserPackages(maintainer: string) {
 
 		packages.push(...body.results.map((result) => result.package));
 
-		if (body.total <= from || body.total > size) {
-			from += size - 1;
-		} else {
+		if (packages.length >= body.total) {
 			return packages;
 		}
 	}

--- a/src/tideliftMeUp.test.ts
+++ b/src/tideliftMeUp.test.ts
@@ -3,13 +3,13 @@ import { describe, expect, it, vi } from "vitest";
 import { createFakePackageData } from "./fakes.js";
 import { tideliftMeUp } from "./tideliftMeUp.js";
 
-const mockNpmUserPackages = vi
+const mockGetNpmUserPackages = vi
 	.fn()
 	.mockResolvedValue([createFakePackageData()]);
 
-vi.mock("npm-user-packages", () => ({
-	get default() {
-		return mockNpmUserPackages;
+vi.mock("./getNpmUserPackages.js", () => ({
+	get getNpmUserPackages() {
+		return mockGetNpmUserPackages;
 	},
 }));
 
@@ -35,22 +35,21 @@ describe("tideliftMeUp", () => {
 	it("calls npmUserPackages with the npm name when the user is logged in", async () => {
 		const username = "abc123";
 
-		mockNpmUserPackages;
 		mockGetNpmWhoami.mockResolvedValue(username);
 
 		await tideliftMeUp();
 
-		expect(mockNpmUserPackages).toHaveBeenCalledWith(username);
+		expect(mockGetNpmUserPackages).toHaveBeenCalledWith(username);
 	});
 
 	it("calls npmUserPackages with the provided username when it exists", async () => {
 		const username = "abc123";
 
-		mockNpmUserPackages.mockResolvedValue([createFakePackageData()]);
+		mockGetNpmUserPackages.mockResolvedValue([createFakePackageData()]);
 		mockGetNpmWhoami.mockRejectedValue(new Error("Should not be called."));
 
 		await tideliftMeUp({ username });
 
-		expect(mockNpmUserPackages).toHaveBeenCalledWith(username);
+		expect(mockGetNpmUserPackages).toHaveBeenCalledWith(username);
 	});
 });

--- a/src/tideliftMeUp.ts
+++ b/src/tideliftMeUp.ts
@@ -1,6 +1,5 @@
-import npmUserPackages from "npm-user-packages";
-
 import { createUserPackagesFilter } from "./createUserPackagesFilter.js";
+import { getNpmUserPackages } from "./getNpmUserPackages.js";
 import { getNpmWhoami } from "./getNpmWhoami.js";
 import { getPackageEstimates } from "./getPackageEstimates.js";
 import { PackageOwnership } from "./types.js";
@@ -21,7 +20,7 @@ export async function tideliftMeUp({
 		throw new Error("Either log in to npm or provide a `username`.");
 	}
 
-	const userPackages = (await npmUserPackages(username)).filter(
+	const userPackages = (await getNpmUserPackages(username)).filter(
 		createUserPackagesFilter({ ownership, since: new Date(since), username })
 	);
 

